filter/series_by_tag.go

```diff
@@ -114,42 +116,95 @@ func createMatcher(spec TagSpec) func(string, map[string]string) bool {
-	patternToMatcher := make(map[string]func(string, map[string]string) bool)
+func NewSeriesByTagPatternIndex(logger moira.Logger, tagSpecsByPattern map[string][]TagSpec) *SeriesByTagPatternIndex {
+	names := make(map[string]bool)
+	nameToPatternMatchers := make(map[string]map[string]MatchingFunc)
+	specialPatternsToMatcher := make(map[string]MatchingFunc)
+
 	for pattern, tagSpecs := range tagSpecsByPattern {
-		matchers := make([]func(string, map[string]string) bool, 0)
+		matchers := make([]MatchingFunc, 0)
+		var nameTagValue string
+		var hasNameTagWithEqualOperationValue bool
+
 		for _, tagSpec := range tagSpecs {
-			matchers = append(matchers, createMatcher(tagSpec))
+			if tagSpec.Name == "name" && tagSpec.Operator == EqualOperator {
+				nameTagValue = tagSpec.Value
+				names[nameTagValue] = true
+				hasNameTagWithEqualOperationValue = true
+			} else {
+				matchers = append(matchers, createMatcher(tagSpec))
+			}
 		}
 
-		patternToMatcher[pattern] = func(metric string, labels map[string]string) bool {
+		matchingFunc := func(metric string, labels map[string]string) bool {
 			for _, matcher := range matchers {
 				if !matcher(metric, labels) {
 					return false
 				}
 			}
 			return true
 		}
+
+		if hasNameTagWithEqualOperationValue {
+			if _, ok := nameToPatternMatchers[nameTagValue]; ok {
+				nameToPatternMatchers[nameTagValue][pattern] = matchingFunc
+			} else {
+				nameToPatternMatchers[nameTagValue] = map[string]MatchingFunc{pattern: matchingFunc}
+			}
+		} else {
+			specialPatternsToMatcher[pattern] = matchingFunc
+		}
 	}
 
-	return &SeriesByTagPatternIndex{patternToMatcher: patternToMatcher}
+	namePatternIndex := NewPatternIndex(logger, getMapKeys(names))
+	return &SeriesByTagPatternIndex{
+		namesPatternIndex:      namePatternIndex,
+		nameToPatternMatchers:  nameToPatternMatchers,
+		specialPatternMatchers: specialPatternsToMatcher}
 }
 
 // MatchPatterns allows to match patterns by metric name and its labels
 func (index *SeriesByTagPatternIndex) MatchPatterns(name string, labels map[string]string) []string {
+	matchers := getMatchingFunctionsFilteredByName(index, name)
 	matchedPatterns := make([]string, 0)
-
-	for pattern, matcher := range index.patternToMatcher {
+	for pattern, matcher := range matchers {
 		if matcher(name, labels) {
 			matchedPatterns = append(matchedPatterns, pattern)
 		}
 	}
-
 	return matchedPatterns
 }
+
+func getMatchingFunctionsFilteredByName(index *SeriesByTagPatternIndex, name string) map[string]MatchingFunc {
+	patternsToMatchers := make(map[string]MatchingFunc)
+	namesPatterns := index.namesPatternIndex.MatchPatterns(name)
+	for _, name := range namesPatterns {
+		for pattern, matchers := range index.nameToPatternMatchers[name] {
+			patternsToMatchers[pattern] = matchers
+		}
+	}
+	for pattern, matchers := range index.specialPatternMatchers {
+		patternsToMatchers[pattern] = matchers
+	}
```

Comment:
Would it make more sense to extract this part outside of this function? There are two reasons for this:
1. `specialPatternMatchers` don't filter patterns by name, so it looks like this is out of scope for current function
2. That way we could exclude useless copying of `specialPatternMatchers` on every function call (especially on large volumes of `specialPatternMatchers`). We can use `specialPatternMatchers` directly in `MatchPatterns` function

Reply:
Done.